PWGHF/HFC/Tasks/taskCorrelationD0.cxx

```diff
@@ -0,0 +1,501 @@
+      bool fillingHFcontainer = false;
+      float invmass = 0;
+      if constexpr (std::is_same_v<hfCandidates, TTracksTrig>) {
+        fillingHFcontainer = true;
+        invmass = InvMassD0(track1);
+        //  TODO: Check how to put this into a Filter
+        if (!isAcceptedCandidate(track1)) {
+          continue;
+        }
+      }
+
+      //  fill single-track distributions
+      target->getTriggerHist()->Fill(CorrelationContainer::kCFStepReconstructed, pt1, multiplicity, posZ, triggerWeight);
+
+      for (auto& track2 : tracks2) {
+
+        if constexpr (std::is_same_v<TTracksAssoc, TTracksTrig>) { // case of h-h correlations where the two types of tracks are the same
+          if (track1 == track2) {
+            continue;
+          }
+        }
+
+        float eta2 = track2.eta();
+        float pt2 = track2.pt();
+        float phi2 = track2.phi();
+        o2::math_utils::bringTo02Pi(phi2);
+
+        //  TODO: add getter for NUE associated efficiency here
+
+        //  TODO: add pair cuts on phi*
+
+        float deltaPhi = phi1 - phi2;
+        //  set range of delta phi in (-pi/2 , 3/2*pi)
+        deltaPhi = RecoDecay::constrainAngle(deltaPhi, -0.5 * PI);
+
+        //  fill pair correlations (HF case needs additional axis for invariant mass)
+        if (!(fillingHFcontainer)) {
+          target->getPairHist()->Fill(CorrelationContainer::kCFStepReconstructed,
+                                      eta1 - eta2, pt2, pt1, multiplicity, deltaPhi, posZ,
+                                      triggerWeight * associatedWeight);
+        } else {
+          target->getPairHist()->Fill(CorrelationContainer::kCFStepReconstructed,
+                                      eta1 - eta2, pt2, pt1, multiplicity, deltaPhi, posZ, invmass,
+                                      triggerWeight * associatedWeight);
+        }
+      }
+    }
+  }
+
+  // =====================================
+  //    process same event correlations
+  // =====================================
+  void processSameRun3(aodCollisions::iterator const& collision,
+                       aodTracks const& tracks,
+                       aod::MFTTracks const& mfttracks,
+                       hfCandidates const& candidates)
+  {
+    if (!(isCollisionSelected(collision, true))) {
+      return;
+    }
+
+    const auto multiplicity = tracks.size();
+    registry.fill(HIST("hMultiplicity"), multiplicity);
+    registry.fill(HIST("hVtxZ"), collision.posZ());
+
+    BinningPolicyBase<2> baseBinning{{axisVertex, axisMultiplicity}, true};
+    int bin = baseBinning.getBin(std::make_tuple(collision.posZ(), multiplicity));
+    registry.fill(HIST("hEventCountSame"), bin);
+
+    sameTPCTPCCh->fillEvent(multiplicity, CorrelationContainer::kCFStepReconstructed);
+    sameTPCMFTCh->fillEvent(multiplicity, CorrelationContainer::kCFStepReconstructed);
+    sameHF->fillEvent(multiplicity, CorrelationContainer::kCFStepReconstructed);
```

Comment:
This you can probably directly move in the following if statements (also in the Run2 process)